### PR TITLE
Add GRPCRequest class for envoy.

### PR DESCRIPTION
**Pull request by chowchow316**

### CODE REVIEW

File: contrib/endpoints/include/api_manager/env_interface.h [hunk 1]
```diff
@@ -58,6 +59,9 @@ class ApiManagerEnvInterface {
   // of the request
   // (possibly before returning).
   virtual void RunHTTPRequest(std::unique_ptr<HTTPRequest> request) = 0;
+
+  // TODO: make it virtual here.
+  void RunGRPCRequest(std::unique_ptr<GRPCRequest> request){};
 };
```
Comment (qiwzhang) [inline contrib/endpoints/include/api_manager/env_interface.h]: = 0;

File: contrib/endpoints/include/api_manager/grpc_request.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  GRPCRequest& set_body(std::string&& value) {
+    body_ = std::move(value);
+    return *this;
+  }
+
+ private:
+  std::function<void(utils::Status, std::string&&)> callback_;
+  std::string method_;
+  // DNS or IP address could be server_ for grpc request.
+  // For envoy, we use cluster name as server_.
```
Comment (qiwzhang) [inline contrib/endpoints/include/api_manager/grpc_request.h]: move the comment to the server() function.

File: contrib/endpoints/include/api_manager/grpc_request.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  // A callback for the environment to invoke when the request is
+  // complete. This will be invoked by the environment exactly once,
+  // and then the environment will drop the std::unique_ptr
+  // referencing the GRPCRequest.
+  void OnComplete(utils::Status status, std::string&& body) {
+    callback_(status, std::move(body));
+  }
+
+  const std::string& method() const { return method_; }
```
Comment (qiwzhang) [inline contrib/endpoints/include/api_manager/grpc_request.h]: add comment 
// the gRPC method to call

File: contrib/endpoints/include/api_manager/grpc_request.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  GRPCRequest& set_server(const std::string& value) {
+    server_ = value;
+    return *this;
+  }
+  GRPCRequest& set_server(std::string&& value) {
+    server_ = std::move(value);
+    return *this;
+  }
+
+  const std::string& service() const { return service_; }
```
Comment (qiwzhang) [inline contrib/endpoints/include/api_manager/grpc_request.h]: // The gRPC service name.

File: contrib/endpoints/include/api_manager/grpc_request.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  GRPCRequest& set_service(const std::string& value) {
+    service_ = value;
+    return *this;
+  }
+  GRPCRequest& set_service(std::string&& value) {
+    service_ = std::move(value);
+    return *this;
+  }
+
+  const std::string& body() const { return body_; }
```
Comment (qiwzhang) [inline contrib/endpoints/include/api_manager/grpc_request.h]: // The request body serialized as string.

File: src/envoy/prototype/api_manager_env.h [hunk 6]
```diff
@@ -26,6 +26,7 @@ class Env : public google::api_manager::ApiManagerEnvInterface,
                      std::function<void()> continuation) override;
   virtual void RunHTTPRequest(
       std::unique_ptr<google::api_manager::HTTPRequest> request) override;
+      std::unique_ptr<google::api_manager::HTTPRequest> request) override;
```
Comment (qiwzhang) [inline src/envoy/prototype/api_manager_env.h]: line 28 and 29 are same?

File: contrib/endpoints/include/api_manager/grpc_request.h (hunk 3 above)
Comment (chowchow316) [inline contrib/endpoints/include/api_manager/grpc_request.h]: Done.

File: contrib/endpoints/include/api_manager/grpc_request.h (hunk 4 above)
Comment (chowchow316) [inline contrib/endpoints/include/api_manager/grpc_request.h]: Done.

File: contrib/endpoints/include/api_manager/grpc_request.h (hunk 5 above)
Comment (chowchow316) [inline contrib/endpoints/include/api_manager/grpc_request.h]: Done.

File: contrib/endpoints/include/api_manager/grpc_request.h (hunk 2 above)
Comment (chowchow316) [inline contrib/endpoints/include/api_manager/grpc_request.h]: Done.

File: src/envoy/prototype/api_manager_env.h (hunk 6 above)
Comment (chowchow316) [inline src/envoy/prototype/api_manager_env.h]: Sorry, added by accident.

File: contrib/endpoints/include/api_manager/env_interface.h (hunk 1 above)
Comment (chowchow316) [inline contrib/endpoints/include/api_manager/env_interface.h]: Done.

File: contrib/endpoints/include/api_manager/grpc_request.h [hunk 7]
```diff
@@ -49,6 +50,8 @@ class GRPCRequest {
     return *this;
   }
 
+  // DNS or IP address could be server_ for grpc request.
```
Comment (qiwzhang) [inline contrib/endpoints/include/api_manager/grpc_request.h]: // DNS or IP address of the gRPC server.

Comment (chowchow316) [inline contrib/endpoints/include/api_manager/grpc_request.h]: Done.

File: contrib/endpoints/include/api_manager/grpc_request.h [hunk 8]
```diff
@@ -50,7 +50,7 @@ class GRPCRequest {
     return *this;
   }
 
-  // DNS or IP address could be server_ for grpc request.
+  // DNS or IP address of the gRPC server.
   // For envoy, we use cluster name as server_.
```
Comment (qiwzhang) [inline contrib/endpoints/include/api_manager/grpc_request.h]: remove second comment line

**Comment by qiwzhang** [comment]
You may need to rebase your change.

**Comment by chowchow316** [comment]
Actually github provides rebase and merge option when submitting PR.

**Pull request merged**